pkg/discovery/queries.go

```diff
@@ -217,101 +167,146 @@ func QueryNSResources(kubeClient kubernetes.Interface, recorder *QueryRecorder,
-	outdir := path.Join(cfg.OutputDir(), ClusterResourceLocation)
-	if len(resources) > 0 {
-		if err := os.MkdirAll(outdir, 0755); err != nil {
-			return errors.WithStack(err)
+		query := func() (time.Duration, error) {
+			return timedListQuery(
+				outdir+"/",
+				groupText+"_"+gvr.Version+"_"+gvr.Resource+".json",
+				lister,
+			)
+		}
+		nsVal := ""
+		if ns != nil {
+			nsVal = *ns
 		}
+
+		timedQuery(recorder, gvr.Resource, nsVal, query)
 	}
 
-	// 2. Execute the non-ns-query
-	for _, resourceKind := range resources {
-		switch resourceKind {
-		case "ServerVersion":
-			objqry := func() (interface{}, error) { return kubeClient.Discovery().ServerVersion() }
-			query := func() (time.Duration, error) {
-				return untypedQuery(cfg.OutputDir(), "serverversion.json", objqry)
+	return nil
+}
+
+func filterResources(gvrs map[schema.GroupVersion][]metav1.APIResource, ns *string, wantResources []string) []schema.GroupVersionResource {
+	results := []schema.GroupVersionResource{}
+	for gv, resources := range gvrs {
+		for _, res := range resources {
+			// Get either namespaced or non-namespaced resources.
+			if (ns != nil && len(*ns) > 0 && !res.Namespaced) ||
+				(ns == nil && res.Namespaced) {
+				if !res.Namespaced {
+					continue
+				}
 			}
-			timedQuery(recorder, "serverversion", "", query)
-			continue
-		case "ServerGroups":
-			objqry := func() (interface{}, error) { return kubeClient.Discovery().ServerGroups() }
-			query := func() (time.Duration, error) {
-				return untypedQuery(cfg.OutputDir(), "servergroups.json", objqry)
+
+			// Double check the resources are listable.
+			listable := false
+			for _, v := range res.Verbs {
+				if v == "list" {
+					listable = true
+					break
+				}
 			}
-			timedQuery(recorder, "servergroups", "", query)
-			continue
-		case "Nodes":
-			// cfg.Nodes configures whether users want to gather the Nodes resource in the
-			// cluster, but we also use that option to guide whether we get node data such
-			// as configz and healthz endpoints.
-
-			// TODO(chuckha) Use a separate configuration like NodeConfiguration for node configz/healthz to make
-			// this switch flow less confusing. "Nodes" is responsible for too much.
-
-			// NOTE: Node data collection is an aggregated time b/c propagating that detail back up
-			// is odd and would pollute some of the output.
-
-			start := time.Now()
-			// TODO(chuckha) look at FieldSelector for list options{}
-			nodeList, err := kubeClient.CoreV1().Nodes().List(metav1.ListOptions{})
-			if err != nil {
-				errlog.LogError(fmt.Errorf("failed to get node list: %v", err))
-				// Do not return or continue because we also want to query nodes as resources
-				break
+			if !listable {
+				continue
 			}
-			nodeNames := make([]string, len(nodeList.Items))
-			for i, node := range nodeList.Items {
-				nodeNames[i] = node.Name
+
+			// Filter for explicit values
+			if len(wantResources) > 0 {
+				if !sliceContains(wantResources, res.Name) {
```

Comment:
This isn't sufficient right now because the list we were using before was all custom code; it didn't necessary correspond to group/kind/resource (usually very close, but may be singular vs plural or casing was different).

Reply:
So the defaults can actually just be removed, though people who run a custom set of resources would then be hit by this.

For maximum backward compatibility, we would need to ensure it works if you pass in an older resource list (I think this whole idea needs some review/design though since there are so many resource types it doesnt make sense for someone to have to specify each and every one they want).

I could add the lower casing in the function but it gets called quite a bit; I might also want to do it when reading in their config. The only thing is that I'm not certain whether or not the name _could_ be upper case, e.g. for CRDs. In which case lower casing the list on load time and then we'd have to lower case the resource name when checking. That starts to get awkward to spread that logic.

Is this a breaking change we want to allow? It avoids inefficiencies of doing case checking or spreading logic into a few different places to solve one problem. Would also only affect people who are running a newer version of sonobuoy with YAML produced by an older version of Sonobuoy. Maybe since this only affects the queries (and no the overall functioning of Sonobuoy or plugins) we can just use a release note to comment on it?

@chuckha curious if you have a thought on this?